src/python/pants/backend/jvm/tasks/scaladoc_gen.py

```diff
@@ -59,9 +64,46 @@ def create_scaladoc_command(self, classpath, gendir, *targets):
 
     args.extend(sources)
 
+    # Check the size of the ScalaDoc command arguments, and use a options file if too long.
+    #
+    # There are a couple practical limits to consider.
+    #
+    # - The length of any individual value must be less than MAX_ARG_STRLEN
+    #   which is hardcoded to 131072 bytes.
+    #
+    # - The total length of the command must be less than the OS's `getconf ARG_MAX`
+    #   value.
+    #
+    #   This value will vary by system. For exmaple, on my MacOS 10.13.6 Twitter-issued
+    #   developer laptop this value is currently 262144. Doing some research on the
+    #   internet shows a wide range of default values for this ranging from 131072 to
+    #   2097152 (2MB). Unfortunately, there doesn't seem to be an easy way to query the
+    #   system for this value. It's not in sysconfig.get_config_vars, or posix.environ,
+    #   or os.environ. We could shell out to `getconf` to get the actual value, but
+    #   anecdotal evidence suggests that there are other layers of limits beyond just
+    #   this POSIX value. POSIX requires ARG_MAX to be at least 4096 bytes, but it's
+    #   extremely unlikely to find a system with that low setting. Rather, I think a
+    #   practical middle ground would be the MAX_ARG_STRLEN fixed value.
+    #
+    #   So, given that rationale, if the arg list is longer than MAX_ARG_STRLEN
+    #   we should put it in an external file, just in case.
+
+    args_joined = ' '.join(map(str, args))
+    if len(args_joined) > MAX_ARG_STRLEN:
+      self.context.log.debug("ScalaDoc arguments too long. Using options file.")
+
+      # dump to file, overwriting any file that already exists
+      options_filepath = os.path.join(self.workdir, 'scaladoc_options.txt')
```

Comment:
In `javadoc_gen.py` we mix in the basename of the directory that the files are being generated for: https://github.com/pantsbuild/pants/blob/4395289966d6ec4a9525367c82629c35fb461b2a/src/python/pants/backend/jvm/tasks/javadoc_gen.py#L71

It's not clear to me why we'd need that (`git blame` doesn't show any rationale), but since both of these tasks are doing the same thing now with argfiles and both subclass `JvmdocGen`, it might be useful to hoist the argfile usage into the base class and have both tasks make use of the same implementation here.

Reply:
That does make sense, though it expands the scope of this change. I would say that sounds like it should be in a separate PR, after this one.

Reply:
Sure! I'll follow up!